File: shared/src/commonMain/kotlin/com/vnidrop/app/core/FileSystemService.kt
package com.vnidrop.app.core

import androidx.compose.runtime.Composable
import uniffi.vnidrop.ReceiveOutputSinkV2

enum class ReceiveFolderKind {
	FileSystemPath,
	/** Shared system Downloads via MediaStore (Android 10+). */
	AndroidPublicDownloads,
	AndroidTreeUri,
}

/** Stable token stored in preferences for [ReceiveFolderKind.AndroidPublicDownloads]. */
const val AndroidPublicDownloadsToken = "media-store:downloads"

data class ReceiveFolder(
	val kind: ReceiveFolderKind,
	val value: String,
	val displayName: String,
)

data class ReceivedStorageInspection(
	val existingBytes: ULong,
	val existingCount: Int,
	val missingCount: Int,
	val inaccessibleCount: Int,
)

enum class FolderAccessStatus {
	Writable,
	PermissionRequired,
	Unavailable,
}

interface FileSystemService {
	val supportsCustomReceiveFolders: Boolean get() = true

	fun defaultReceiveFolder(): ReceiveFolder
	fun effectiveReceiveFolder(configuredFolder: ReceiveFolder): ReceiveFolder =
		if (supportsCustomReceiveFolders) configuredFolder else defaultReceiveFolder()
	suspend fun validateReceiveFolder(folder: ReceiveFolder): FolderAccessStatus
	suspend fun inspectReceivedArtifacts(artifacts: List<ReceivedArtifactModel>): ReceivedStorageInspection
	suspend fun temporaryUsage(receiveFolder: ReceiveFolder): ULong
	fun createReceiveOutputSink(folder: ReceiveFolder): ReceiveOutputSinkV2?
	fun canRevealReceiveFolder(folder: ReceiveFolder): Boolean = false
	suspend fun revealReceiveFolder(folder: ReceiveFolder): Result<Unit> =
		Result.failure(UnsupportedOperationException("Revealing the receive folder is not supported"))
	/** Releases only app-owned picker copies; implementations must never delete original user sources. */
	suspend fun discardPickedFiles(files: List<PickedShareFile>) = Unit
	suspend fun sharePickedFile(
		repository: CoreGateway,
		file: PickedShareFile,
		transferName: String,
		senderName: String,
		accessPolicy: ShareAccessPolicy,
	): Result<Share> = sharePickedFiles(repository, listOf(file), transferName, senderName, accessPolicy)

	suspend fun sharePickedFiles(
		repository: CoreGateway,
		files: List<PickedShareFile>,
		transferName: String,
		senderName: String,
		accessPolicy: ShareAccessPolicy,
	): Result<Share>
}

@Composable
expect fun rememberFileSystemService(): FileSystemService

fun ReceiveFolder.isFileSystemPath(): Boolean =
	kind == ReceiveFolderKind.FileSystemPath

File: apple/VniDrop/Core/FileSystemService.swift
import Foundation
import VnidropCore

/// A file/folder selected for sharing, ported from `PickedShareFile` in
/// `core/FilePicker.kt`.
struct PickedShareFile: Equatable, Identifiable, Sendable {
	let value: String
	let displayName: String
	var sizeBytes: UInt64? = nil
	var thumbnailData: Data? = nil
	/// App-owned picker copy that may be deleted after import or abandonment.
	var isTemporaryCopy: Bool = false
	/// When true, `value` is a directory (path or security-scoped folder URL).
	var isDirectory: Bool = false

	var id: String { value }
}

/// Receive-destination and share-source platform bridge, ported from
/// `core/FileSystemService.kt` and its iOS/desktop actuals.
@MainActor
protocol FileSystemService {
	var supportsCustomReceiveFolders: Bool { get }

	func defaultReceiveFolder() -> ReceiveFolder
	func effectiveReceiveFolder(_ configured: ReceiveFolder) -> ReceiveFolder
	func validateReceiveFolder(_ folder: ReceiveFolder) async -> FolderAccessStatus
	func canRevealReceiveFolder(_ folder: ReceiveFolder) -> Bool
	func revealReceiveFolder(_ folder: ReceiveFolder) async -> Result<Void, Error>
	/// Releases only app-owned picker copies; never deletes original user sources.
	func discardPickedFiles(_ files: [PickedShareFile]) async
	func sharePickedFiles(
		repository: CoreRepository,
		files: [PickedShareFile],
		transferName: String,
		senderName: String,
		accessPolicy: ShareAccessPolicy
	) async -> Result<Share, Error>
}

extension FileSystemService {
	var supportsCustomReceiveFolders: Bool { true }

	func effectiveReceiveFolder(_ configured: ReceiveFolder) -> ReceiveFolder {
		supportsCustomReceiveFolders ? configured : defaultReceiveFolder()
	}

	func canRevealReceiveFolder(_ folder: ReceiveFolder) -> Bool { false }

	func revealReceiveFolder(_ folder: ReceiveFolder) async -> Result<Void, Error> {
		.failure(InvitationError.message("Revealing the receive folder is not supported"))
	}

	func discardPickedFiles(_ files: [PickedShareFile]) async {}
}

extension ReceiveFolder {
	var isFileSystemPath: Bool { kind == .fileSystemPath }
}
